onnx/defs/tensor/defs.cc

```diff
@@ -49,6 +49,36 @@ from the shape argument.)DOC")
     .TypeConstraint("T", { "tensor(float16)", "tensor(float)", "tensor(double)" },
 	            "Constrain input and output types to float tensors.");
 
+ONNX_OPERATOR_SCHEMA(Shape)
+    .SetDoc(R"DOC(
+Takes a tensor as input and outputs an 1D int64 tensor with the shape of the input tensor.
+)DOC")
+    .Input(0, "data", "An input tensor.", "T")
+    .Output(0, "shape", "Shape of the input tensor", "T1")
+    .TypeConstraint("T", { "tensor(float16)",  "tensor(float)", "tensor(double)",
+                           "tensor(int8)", "tensor(int16)",  "tensor(int32)", "tensor(int64)",
+                           "tensor(uint8)", "tensor(uint16)",
+                           "tensor(bool)" },
+                "Input tensor can be of arbitrary type.")
+    .TypeConstraint("T1", { "tensor(int64)"},
+	            "Constrains output to int64 tensor.");
+
+
+ONNX_OPERATOR_SCHEMA(Size)
```

Comment:
Shape + ReduceSum?

Reply:
@ebarsoum I thought about that, but 1) Right now onnx doesn't have Prod op (there is ReduceProd, but it's a little bit overkill to use in this case IMO) 2) we in general want to follow the numpy api and it has both `shape` and `size` (pytorch and caffe2 as well).

Reply:
@bddppq we can add Prod OP, the problem is that how generic `size` is? For example, for batch of variable length sequences do I want to get the total number of sample per each sequence multiply by the batch size? In some cases, you would want the number of element per sample and in others you want across the batch axis. Regarding the availability of size in numpy, Caffe2 and PyTorch. This is API level not IR level, so that shouldn't impact the inclusion in the ONNX, especially if we can represented in ONNX.

How `size` in PyTorch and Caffe 2 work with variable length sequences?

Reply:
@ebarsoum 
1. I agree we should add Prod, it's a fundamental math operation.
2. What do you mean by "variable" length sequences? Are you saying that each sequence could have different length?
3. IMO size is a pretty bare bone op that shouldn't take batching into consideration. numpy, caffe2, pytorch, mxnet and tensorflow return the total number.

Considering we can easily add an optimization pass right now, I don't have strong preference between size or shape + prod (although slightly prefer the former which you can tell from my arguments).

@dzhulgakov @jamesr66a thoughts?

Reply:
(2) yes.
Shape and Size usually used to compute some form of normalization, Shape is generic enough and we need it.